crates/automations/src/executors.rs

```diff
@@ -0,0 +1,306 @@
+    let _instant = heartbeat_ticks.tick().await; // Discard immediate first tick.
+
+    // Build a Future which forever maintains our heartbeat or fails.
+    let update_heartbeats = async {
+        loop {
+            let _instant = heartbeat_ticks.tick().await;
+
+            last_heartbeat =
+                match update_heartbeat(&pool, task_id, heartbeat_timeout, last_heartbeat).await {
+                    Ok(last_heartbeat) => last_heartbeat,
+                    Err(err) => return err,
+                }
+        }
+    };
+    tokio::pin!(update_heartbeats);
+
+    // Poll `executor` and `update_heartbeats` in tandem, so that a failure
+    // to update our heartbeat also cancels the executor.
+    let outcome = tokio::select! {
+        outcome = executor.poll(task_id, parent_id, &mut state, &mut inbox) => { outcome? },
+        err = &mut update_heartbeats => return Err(err),
+    };
+
+    // The possibly long-lived polling operation is now complete.
+    // Build a Future that commits a (hopefully) brief transaction of `outcome`.
+    let persist_outcome = async {
+        let mut txn = pool.begin().await?;
+        () = persist_outcome(outcome, &mut *txn, task_id, parent_id, state, inbox).await?;
+        Ok(txn.commit().await?)
+    };
+
+    // Poll `persist_outcome` while continuing to poll `update_heartbeats`,
+    // to guarantee we cannot commit an outcome after our lease is lost.
+    tokio::select! {
+        result = persist_outcome => result,
+        err = update_heartbeats => Err(err),
+    }
+}
+
+async fn update_heartbeat(
+    pool: &sqlx::PgPool,
+    task_id: models::Id,
+    heartbeat_timeout: std::time::Duration,
+    expect_heartbeat: String,
+) -> anyhow::Result<String> {
+    let update = sqlx::query!(
+        r#"
+        UPDATE internal.tasks
+        SET heartbeat = NOW()
+        WHERE task_id = $1 AND heartbeat::TEXT = $2
+        RETURNING heartbeat::TEXT AS "heartbeat!";
+        "#,
+        task_id as models::Id,
+        expect_heartbeat,
+    )
+    .fetch_optional(pool);
+
+    // We must guard against both explicit errors and also timeouts when updating
+    // the heartbeat, to ensure we bubble up an error that cancels our paired
+    // executor prior to `heartbeat_timeout` elapsing.
+    let updated = match tokio::time::timeout(heartbeat_timeout / 4, update).await {
+        Ok(Ok(Some(updated))) => updated,
+        Ok(Ok(None)) => anyhow::bail!("task heartbeat was unexpectedly updated externally"),
+        Ok(Err(err)) => return Err(anyhow::anyhow!(err).context("failed to update task heartbeat")),
+        Err(err) => return Err(anyhow::anyhow!(err).context("timed out updating task heartbeat")),
+    };
+
+    tracing::info!(
+        last = expect_heartbeat,
+        next = updated.heartbeat,
+        "updated task heartbeat"
+    );
+
+    Ok(updated.heartbeat)
+}
+
+async fn persist_outcome(
+    outcome: PollOutcome<BoxedRaw>,
+    txn: &mut sqlx::PgConnection,
```

Comment:
This is a transaction -- it's just been downcast to the PgConnection. I did that because in newer versions of sqlx, transaction is not an executor and can't be used with sqlx::query.

We definitely want full transaction semantics around task management!